applications/GeoMechanicsApplication/custom_processes/geo_extrapolate_integration_point_values_to_nodes_process.cpp

```diff
@@ -0,0 +1,228 @@
+{
+    InitializeAverageVariablesForElements();
+    InitializeVectorAndMatrixZeros();
+}
+
+void GeoExtrapolateIntegrationPointValuesToNodesProcess::InitializeAverageVariablesForElements() const
+{
+    VariableUtils().SetNonHistoricalVariable(mrAverageVariable, 0.0, mrModelPart.Nodes());
+
+    block_for_each(mrModelPart.Elements(), [this](Element& rElement) {
+        if (rElement.IsActive()) {
+            for (auto& node : rElement.GetGeometry()) {
+                auto& node_var_to_update = node.GetValue(mrAverageVariable);
+                AtomicAdd(node_var_to_update, 1.0);
+            }
+        }
+    });
+
+    mrModelPart.GetCommunicator().AssembleNonHistoricalData(mrAverageVariable);
+}
+
+void GeoExtrapolateIntegrationPointValuesToNodesProcess::InitializeVectorAndMatrixZeros()
+{
+    if (mrModelPart.Elements().empty()) return;
+
+    auto first_element = mrModelPart.Elements().begin();
+
+    const auto integration_points_number =
+        first_element->GetGeometry().IntegrationPointsNumber(first_element->GetIntegrationMethod());
+
+    const ProcessInfo& r_process_info = mrModelPart.GetProcessInfo();
+    InitializeZerosOfVectorVariables(*first_element, integration_points_number, r_process_info);
+    InitializeZerosOfMatrixVariables(*first_element, integration_points_number, r_process_info);
+}
+
+void GeoExtrapolateIntegrationPointValuesToNodesProcess::InitializeZerosOfVectorVariables(
+    Element& rFirstElement, SizeType NumberOfIntegrationPoints, const ProcessInfo& rProcessInfo)
+{
+    for (const auto p_var : mVectorVariables) {
+        std::vector<Vector> values_on_integration_points(NumberOfIntegrationPoints);
+        rFirstElement.CalculateOnIntegrationPoints(*p_var, values_on_integration_points, rProcessInfo);
+        mZeroValuesOfVectorVariables.try_emplace(p_var, ZeroVector(values_on_integration_points[0].size()));
+    }
+}
+
+void GeoExtrapolateIntegrationPointValuesToNodesProcess::InitializeZerosOfMatrixVariables(
+    Element& rFirstElement, SizeType NumberOfIntegrationPoints, const ProcessInfo& rProcessInfo)
+{
+    for (const auto p_var : mMatrixVariables) {
+        std::vector<Matrix> values_on_integration_points(NumberOfIntegrationPoints);
+        rFirstElement.CalculateOnIntegrationPoints(*p_var, values_on_integration_points, rProcessInfo);
+        mZeroValuesOfMatrixVariables.try_emplace(
+            p_var, ZeroMatrix(values_on_integration_points[0].size1(),
+                              values_on_integration_points[0].size2()));
+    }
+}
+
+void GeoExtrapolateIntegrationPointValuesToNodesProcess::ExecuteFinalizeSolutionStep()
+{
+    InitializeVariables();
+
+    block_for_each(mrModelPart.Elements(), [this](Element& rElem) {
+        if (rElem.IsActive()) {
+            const Matrix extrapolation_matrix = GetExtrapolationMatrix(rElem);
+            AddIntegrationContributionsForAllVariableLists(rElem, extrapolation_matrix);
```

Comment:
Avoid making an unnecessary copy:
```suggestion
            AddIntegrationContributionsForAllVariableLists(rElem, GetExtrapolationMatrix(rElem));
```

Reply:
Good catch!